viewer/src/culling/ByVisibilityGpuSectorCuller.ts

```diff
@@ -0,0 +1,248 @@
+ * and loads sectors based on priority within a budget.
+ */
+export class ByVisibilityGpuSectorCuller implements SectorCuller {
+  public static readonly DefaultCostLimit = 50 * 1024 * 1024;
+  public static readonly DefaultHighDetailProximityThreshold = 10;
+
+  private readonly options: Required<ByVisibilityGpuSectorCullerOptions>;
+  private readonly camera: THREE.PerspectiveCamera;
+  private readonly models: CadModel[] = [];
+  private readonly lastUpdate: {
+    matrix: THREE.Matrix4;
+    projectionMatrix: THREE.Matrix4;
+    takenSectors: TakenSectorMap;
+  };
+
+  constructor(camera: THREE.PerspectiveCamera, options?: ByVisibilityGpuSectorCullerOptions) {
+    this.options = {
+      costLimit: options && options.costLimit ? options.costLimit : ByVisibilityGpuSectorCuller.DefaultCostLimit,
+      determineSectorCost:
+        options && options.determineSectorCost ? options.determineSectorCost : computeSectorCostAsDownloadSize,
+      highDetailProximityThreshold:
+        options && options.highDetailProximityThreshold
+          ? options.highDetailProximityThreshold
+          : ByVisibilityGpuSectorCuller.DefaultHighDetailProximityThreshold,
+
+      coverageUtil: options && options.coverageUtil ? options.coverageUtil : new GpuOrderSectorsByVisibilityCoverage()
+    };
+    this.lastUpdate = {
+      matrix: new THREE.Matrix4().fromArray(invalidMatrixArray),
+      projectionMatrix: new THREE.Matrix4().fromArray(invalidMatrixArray),
+      takenSectors: new TakenSectorMap(this.options.determineSectorCost)
+    };
+    this.camera = camera;
+  }
+
+  addModel(cadModel: CadModel) {
+    const { coverageUtil } = this.options;
+    this.models.push(cadModel);
+    coverageUtil.addModel(cadModel.scene, cadModel.modelTransformation);
+  }
+
+  determineSectors(input: DetermineSectorsByProximityInput): WantedSector[] {
+    const takenSectors = this.update();
+
+    const wantedForScene = takenSectors.collectWantedSectors(input.sectorScene);
+    console.log(
+      `Scene: ${wantedForScene.length} (${wantedForScene.filter(x => !Number.isFinite(x.priority)).length} required)`
+    );
+
+    return wantedForScene;
+  }
+
+  get lastWantedSectors(): PrioritizedWantedSector[] {
+    return this.models.flatMap(m => this.lastUpdate.takenSectors.collectWantedSectors(m.scene));
+  }
+
+  private update(): TakenSectorMap {
+    const { coverageUtil } = this.options;
+    const takenSectors = this.lastUpdate.takenSectors;
+    // Necessary to update?
+    const changed =
+      !this.lastUpdate.matrix.equals(this.camera.matrix) ||
+      !this.lastUpdate.projectionMatrix.equals(this.camera.projectionMatrix);
+
+    if (!changed) {
+      return takenSectors;
+    }
+    takenSectors.clear();
+    this.models.forEach(x => takenSectors.initializeScene(x.scene));
+
+    // Update wanted sectors
+    const prioritized = coverageUtil.orderSectorsByVisibility(this.camera);
+    this.lastUpdate.matrix = this.camera.matrix.clone();
```

Comment:
Use `this.lastupdate.matrix.copy(this.camera.matrix)` instead to avoid the extra allocation.

Reply:
Good catch, fixed.